# Health Data Science CDT - Tutorial: Pathogen sequencing 

## Detecting antimicrobial resistance determinants

### Background
As discussed in the lecture, antimicrobial resistance can be determined by several mechanisms
* Modification of the target
* Degradation of the antimicrobial
* Reduced influx or increased efflux

In turn each mechanism can be brought about by several changes at the genome level
* Single nucleotide polymorphisms (SNPs) in genes or their regulators
* Small insertions or deletions - indels in genes or their regulators
* Gain of new genes on mobile genetic elements including plasmids

Three main apporaches can be used to look for antimicrobial resistance determinants
* Reference based mapping - good for SNPs and some indels, but can't detect novel genetic material like gain of a plasmid
* De novo assembly - good for gain of novel material and larger insertions and deletions

Neither approach works well where there are multiple copies of a gene in a genome - we will develop a bespoke approach for this.

### Overview
In this tutorial you will learn how to use various scriptable approaches to detect different antimicrobial resistance determinants. The tutorial will be mostly based using python, but also rely on additional software packages.

### The world's worst super gonorrhoea
In February 2018, a patient was diagnosed with the most difficult to treat gonorrhoea ever described. You can read more about it here - https://www.bbc.co.uk/news/health-43571120 - and his suscessful treatment here - https://www.bbc.co.uk/news/health-43840505. The more scientific version is here - https://www.eurosurveillance.org/content/10.2807/1560-7917.ES.2018.23.27.1800323 - but I would encourage not to look at this too closely to start with, as it contains many of the answers!

Two isolates from this infection were sequenced using Illumina technology. You are provided with
* Sequence reads - the raw output of the sequencer
* A ready-made mapped concensus fasta file
* A ready-made de novo assembly

The mapped file was generated by mapping reads to a reference genome (e.g. using BWA mem), and searching these for variants from the reference genome (e.g. using samtools mpileup), and quality filtering them to identify variants of high quality. The de novo assembly was made an assembler (e.g. SPAdes).

### Suggested resources
You will need to use the following biopython tools in the tutorial
* https://biopython.org/wiki/SeqIO - for reading in the mapped files
* http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc93 - for searching de novo assemblies

You will also need a way to re-map the raw reads to a novel reference
* You can do this with BWA mem - https://github.com/lh3/bwa
* You will need a way to analyse the mapped files - I suggest using pysam to run samtools - https://pysam.readthedocs.io/en/latest/api.html#introduction

Most of the relevant resistance mutations and variants in Neisseria gonorrhoeae (the bacteria that causes gonorrhoea) are catelogued in the NG-STAR database - https://ngstar.canada.ca/alleles/query?lang=en


