{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1 - Explain ciprofloxacin resistance\n",
    "\n",
    "**A common mutation causing ciprofloxacin resistance is found in the *gyrA* and results in the following amino acid change - S91F. Can you confirm if this is present in the sequence data, using the mapped data, and find another resistance determinant?**\n",
    "\n",
    "This relies on using python to identify the gene in the whole genome andconvert DNA sequence to protein sequences, so we can then compare the reference genome with our genome of interest to look for differences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 1\n",
    "\n",
    "Find the location of the *gyrA* gene in the reference file using the NCBI website - https://www.ncbi.nlm.nih.gov/nuccore/NC_011035.1?report=graph - use the find function on the webpage\n",
    "\n",
    "- Coordinates of the *gyrA* gene on [NCBI](https://www.ncbi.nlm.nih.gov/nuccore/NC_011035.1?report=graph): 1051396 to 1054146\n",
    "- In Python: 1051395 to 1054145.\n",
    "\n",
    "![NCBI Screenshot](task1_tip1.png)\n",
    "\n",
    "### Tip 2\n",
    "\n",
    "Use biopython's `SeqIO.read()` function to read in the mapped fasta file `cdt-tutorial/super_gonorrhoea/super_gc_mapped.fa` (https://biopython.org/wiki/SeqIO)\n",
    "\n",
    "**Change path.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "super_gc_mapped_fa = SeqIO.read('../super_gonorrhoea/super_gc_mapped.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 3\n",
    "\n",
    "You can will need to extract the DNA sequence of the gene using it's coordinates within the whole genome (remember that these number from zero in python, but from one on the NCBI website!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: OxfordGC_R00000419_R00000419\n",
      "Name: OxfordGC_R00000419_R00000419\n",
      "Description: OxfordGC_R00000419_R00000419 Software:/home/compass/PIPELINE/mmmPipeline/compass/g4_basecall.py Version:1.0.1\n",
      "Number of features: 0\n",
      "Seq('ATGACCGACGCAACCATCCGCCACGACCACAAATTCGCCCTCGAAACCCTGCCC...TGA', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "super_gc_GyrA_extract = super_gc_mapped_fa[1051395:1054146]\n",
    "print(super_gc_GyrA_extract)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 4\n",
    "\n",
    "You will need to translate the DNA sequences to a protein sequence, `seq.translate()` in biopython (for *gyrA* the gene is in the same direction as the DNA is numbered, if it were not you would need to generate the reverse complement sequence first), if this works you should end up with a string of amino acids represented as single letters and ending with a stop codon shown as an asterisk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('MTDATIRHDHKFALETLPVSLEDEMRKSYLDYAMSVIVGRALPDVRDGLKPVHR...EN*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n"
     ]
    }
   ],
   "source": [
    "super_gc_GyrA = super_gc_GyrA_extract.translate()\n",
    "print(super_gc_GyrA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 5\n",
    "\n",
    "Compare the amino acid sequences from the reference (`cdt-tutorial/super_gonorrhoea/reference.fa` ) with the sequence from the case using python, does the reference share any mutations or have any different mutations in *gyrA* (hint the reference sequence is also ciprofloxacin resistant)?\n",
    "\n",
    "**Change path.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'F'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore 90-th position\n",
    "super_gc_GyrA[90]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 'F'on the 91st (90th from 0) place, so we confirm the amino acid change - S91F.\n",
    "\n",
    "**S91F** means S->F at position 91. Indeed, we have 'F' at position 91. 'S' (Serine) and 'F' (Phenylalanine), in this case, are amino-acids in their [one-letter codes](https://en.wikipedia.org/wiki/Amino_acid#:~:text=Amino%20acids%20are%20the%20structural,to%20two%20neighboring%20amino%20acids.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('MTDATIRHDHKFALETLPVSLEDEMRKSYLDYAMSVIVGRALPDVRDGLKPVHR...EN*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n"
     ]
    }
   ],
   "source": [
    "# reference\n",
    "\n",
    "reference_mapped_fa = SeqIO.read('../super_gonorrhoea/reference.fa', 'fasta')\n",
    "reference_GyrA = reference_gc_mapped_fa[1051395:1054146].translate()\n",
    "print(reference_GyrA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 inconsistent index(es)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([95]), array(['A'], dtype='<U1'), array(['G'], dtype='<U1'))"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "inconsistent_idxs = np.array((reference_GyrA)) != np.array((super_gc_GyrA))\n",
    "print(sum(inconsistent_idxs), 'inconsistent index(es)')\n",
    "\n",
    "(np.arange(1, len(reference_GyrA)+1)[inconsistent_idxs],\n",
    "np.array(super_gc_GyrA)[inconsistent_idxs],\n",
    "np.array(reference_GyrA)[inconsistent_idxs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "95 A G\n"
     ]
    }
   ],
   "source": [
    "for i, (aa1, aa2) in enumerate(zip(super_gc_GyrA, reference_GyrA)):\n",
    "    if aa1!=aa2:\n",
    "        print(i+1, aa1, aa2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2 - Explain the tetracycline resistance\n",
    "**Tetracycline resistance is conferred by gain of a new gene, *tetM*, carried on a plasmid. Is this gene present?**  \n",
    "To do this we will use a tool called BLAST. This is an optimised algorithm for searching for one or more sequences (the query) within a database of sequences. In this example we will treat the gene we are trying to find as the query and the database will be the contigs that make up the assembly of our genome of interest.\n",
    "\n",
    "I suggest you read the Process and Algorithm sections of the wikipedia article on BLAST before continuing - as this will help with parsing the output generated - https://en.wikipedia.org/wiki/BLAST_(biotechnology)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip1\n",
    "Use the de novo assembly provided `super_gc_contigs.fa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 2\n",
    "You will need to create blast database files for the de novo assembly with this command:  \n",
    "```makeblastdb -dbtype nucl -in super_gc_contigs.fa```\n",
    "\n",
    "This can be done using this snippet in python to run the command within the notebook:\n",
    "```\n",
    "import os\n",
    "cmd = 'makeblastdb -dbtype nucl -in cdt-tutorial/super_gonorrhoea/super_gc_contigs.fa'\n",
    "os.system(cmd)\n",
    "```\n",
    "\n",
    "**Change path.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "cmd = 'makeblastdb -dbtype nucl -in ../super_gonorrhoea/super_gc_contigs.fa'\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['super_gc_contigs.fa.ntf',\n",
       " 'tetM.fa',\n",
       " 'super_gc_contigs.fa.ndb',\n",
       " 'penA_alleles.fa',\n",
       " 'super_gc_contigs.fa.nsq',\n",
       " 'super_gc_contigs.fa.nin',\n",
       " 'reference.fa',\n",
       " 'super_gc_mapped.fa',\n",
       " '23s_reference.fa',\n",
       " 'super_gc_contigs.fa',\n",
       " 'super_gc_contigs.fa.nto',\n",
       " 'super_gc_contigs.fa.nhr',\n",
       " 'super_gc_contigs.fa.not']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check that we generated `super_gc_contigs.fa.n*` files\n",
    "os.listdir('../super_gonorrhoea/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 3\n",
    "Use the biopython blast module to run a blast command that searches for the tetM gene, to get you started:  \n",
    "\n",
    "```\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from io import StringIO\n",
    "\n",
    "os.chdir('cdt-tutorial/super_gonorrhoea/')\n",
    "\n",
    "geneFile = 'tetM.fa'\n",
    "assemblyFile = 'super_gc_contigs.fa'\n",
    "\n",
    "cline = NcbiblastnCommandline( query=geneFile, \n",
    "         db=assemblyFile, \n",
    "         evalue=0.01, \n",
    "         outfmt=5\n",
    "         )\n",
    "stdout, stderr = cline()\n",
    "fp = StringIO( stdout )\n",
    "blast_records = [r for r in NCBIXML.parse( fp )]\n",
    "```\n",
    "This snippet will run blast, saving the output in XML format, this is then passed back to python. You should be able interrogate the `blast_records` object to look for any matches found. Details of the contents of the object can be found here - http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc93."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from io import StringIO\n",
    "\n",
    "os.chdir('../super_gonorrhoea/')\n",
    "\n",
    "geneFile = 'tetM.fa'\n",
    "assemblyFile = 'super_gc_contigs.fa'\n",
    "\n",
    "cline = NcbiblastnCommandline( query=geneFile, \n",
    "         db=assemblyFile, \n",
    "         evalue=0.01, \n",
    "         outfmt=5\n",
    "         )\n",
    "stdout, stderr = cline()\n",
    "fp = StringIO( stdout )\n",
    "blast_records = [r for r in NCBIXML.parse( fp )]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 4\n",
    "- Did you find an **alignment match**?\n",
    "- How much of the total length of the tetM gene was matched?\n",
    "- How similar was it to the tetM sequence in `tetM.fa`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tetM has 1 alignments\n",
      "\n",
      "tetM length: 1932\n",
      "Alignment length: 1932\n",
      "Number of identities: 1932\n",
      "Number of gaps: 0\n"
     ]
    }
   ],
   "source": [
    "# from answers\n",
    "for record in blast_records:\n",
    "    print('%s has %s alignments\\n'%(record.query, len(record.alignments)))\n",
    "    if len(record.alignments)>0:\n",
    "        hits = []\n",
    "        for alignment in record.alignments:\n",
    "            for hsp in alignment.hsps:\n",
    "                print('tetM length: %s'%len(hsp.query))\n",
    "                print('Alignment length: %s'%hsp.align_length)\n",
    "                print('Number of identities: %s'%hsp.identities)\n",
    "                print('Number of gaps: %s'%hsp.gaps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Full match.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3 - Explain the resistance to ceftriaxone\n",
    "**Here we build on the example above to look for the best match to the *penA* gene using BLAST.**  \n",
    "A number of mechanisms can change how susceptible N. gonorrhoeae is to ceftriaxone, however the gene *penA* is the most important. Although a few SNPs within the *penA* gene, e.g. A311V, T316P and T483S, the rest of the gene is very diverse (it was generated by recombination events with similar genes in other Neisseria species), and so mapping approaches do not work well for finding these variants - as the sequence is too different to the reference to map reliably. We therefore use a de novo assembly and a database of know *penA* sequences or alleles. We will compare our sequence of interest to this list of alleles. \n",
    "\n",
    "### Tip 1\n",
    "Start as in Task 2, using the assembly as the blast database and this time using the list of *penA* alleles as the query `penA_alleles.fa` (if you want you can also try this task in reverse using the list of *penA* alleles as the database, to do this you'll need to run makeblastdb on the command line as above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneFile = 'penA_alleles.fa'\n",
    "assemblyFile = 'super_gc_contigs.fa'\n",
    "\n",
    "cline = NcbiblastnCommandline( query=geneFile, \n",
    "         db=assemblyFile, \n",
    "         evalue=0.01, \n",
    "         outfmt=5\n",
    "         )\n",
    "stdout, stderr = cline()\n",
    "fp = StringIO( stdout )\n",
    "blast_records = [r for r in NCBIXML.parse( fp )]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Print the hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "penA_1.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.009 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.01 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_1.011 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1582\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.009 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.01 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.011 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.012 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.013 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.014 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.015 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.016 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1575\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.017 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.018 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.019 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.02 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.021 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1575\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.022 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.023 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.024 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.025 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.026 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.027 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.028 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.029 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.03 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.031 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.032 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_2.033 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_3.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1588\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_3.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1589\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_3.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_3.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1588\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_4.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_5.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1587\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_5.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1587\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_5.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1588\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_5.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_5.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_5.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_7.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_9.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_10.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1676\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1683\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.003 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1674\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.004 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1675\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.005 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1675\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.006 has 1 alignments\n",
      "penA length: 1746\n",
      "Alignment length: 1746\n",
      "Number of identities: 1673\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_10.007 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1684\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_11.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_12.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_13.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_13.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1577\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_13.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_13.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1576\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.009 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.01 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_14.011 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_15.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.002 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1584\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.003 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.004 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.005 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.006 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.007 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_15.008 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_16.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_17.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_18.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_18.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_18.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1593\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.009 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_19.01 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_21.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_21.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_21.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1596\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.005 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1593\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.006 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.007 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.008 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.009 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.01 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.011 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.012 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_22.013 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_27.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1680\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_27.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1683\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_27.003 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1676\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_34.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1664\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.003 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.004 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.005 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.006 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1665\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.007 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.008 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1665\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.009 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.01 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.011 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.013 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.014 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.015 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1664\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.016 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.017 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1661\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.018 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_34.019 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1662\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_35.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1613\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_35.002 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1611\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_35.003 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1614\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_37.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1661\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_38.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1590\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_39.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1662\n",
      "Number of gaps: 10\n",
      "\n",
      "penA_40.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_41.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1563\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_41.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1560\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_42.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_43.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_43.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_43.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_43.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_44.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_44.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_44.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_44.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_45.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1595\n",
      "Number of gaps: 8\n",
      "\n",
      "penA_46.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_47.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1690\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_48.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_49.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_50.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_51.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1657\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_52.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1655\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_52.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_53.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1660\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_53.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1659\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_54.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_55.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1661\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_56.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1569\n",
      "Number of gaps: 15\n",
      "\n",
      "penA_57.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1576\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_58.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1661\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_59.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1651\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_60.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1749\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_61.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1577\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_62.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1590\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_63.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1669\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_64.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1707\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_65.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_66.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1654\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_67.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1672\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_67.002 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1666\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_68.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_69.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1587\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_69.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_69.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_70.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_71.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1689\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_72.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_73.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1613\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_74.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1613\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_75.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1666\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_76.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1570\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_76.002 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1569\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_77.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_78.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1663\n",
      "Number of gaps: 9\n",
      "\n",
      "penA_78.002 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1658\n",
      "Number of gaps: 9\n",
      "\n",
      "penA_79.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1566\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_79.002 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1569\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_80.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1530\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_81.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1624\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_82.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1590\n",
      "Number of gaps: 8\n",
      "\n",
      "penA_83.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_84.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1590\n",
      "Number of gaps: 8\n",
      "\n",
      "penA_85.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1663\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_86.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1587\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_87.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_87.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_88.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1589\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_89.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1609\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_90.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_91.001 has 1 alignments\n",
      "penA length: 1754\n",
      "Alignment length: 1754\n",
      "Number of identities: 1685\n",
      "Number of gaps: 10\n",
      "\n",
      "penA_92.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1671\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_93.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1643\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_93.002 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1643\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_94.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1605\n",
      "Number of gaps: 8\n",
      "\n",
      "penA_95.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_96.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_97.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_98.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_98.002 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_99.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_100.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.002 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.003 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.004 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.005 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1584\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.006 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.007 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.008 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1583\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.009 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.01 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.011 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1580\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.012 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.013 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.014 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_100.015 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_101.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1703\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_102.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_103.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1591\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_103.002 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_103.003 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1590\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_103.004 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1591\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_104.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_105.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1552\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_106.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1581\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_107.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1577\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_108.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1690\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_109.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_110.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1643\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_111.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1640\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_112.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1632\n",
      "Number of gaps: 9\n",
      "\n",
      "penA_113.001 has 1 alignments\n",
      "penA length: 1752\n",
      "Alignment length: 1752\n",
      "Number of identities: 1612\n",
      "Number of gaps: 6\n",
      "\n",
      "penA_114.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1630\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_115.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1642\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_116.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1637\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_117.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1652\n",
      "Number of gaps: 2\n",
      "\n",
      "penA_118.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1584\n",
      "Number of gaps: 4\n",
      "\n",
      "penA_119.001 has 1 alignments\n",
      "penA length: 1753\n",
      "Alignment length: 1753\n",
      "Number of identities: 1582\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_120.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1583\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_121.001 has 1 alignments\n",
      "penA length: 1713\n",
      "Alignment length: 1713\n",
      "Number of identities: 1579\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_121.002 has 1 alignments\n",
      "penA length: 1713\n",
      "Alignment length: 1713\n",
      "Number of identities: 1577\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_122.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1587\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_123.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1584\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_124.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1677\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_125.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1580\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_126.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_127.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1647\n",
      "Number of gaps: 12\n",
      "\n",
      "penA_128.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1557\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_129.001 has 1 alignments\n",
      "penA length: 1750\n",
      "Alignment length: 1750\n",
      "Number of identities: 1629\n",
      "Number of gaps: 5\n",
      "\n",
      "penA_130.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1659\n",
      "Number of gaps: 9\n",
      "\n",
      "penA_131.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1680\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_132.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1586\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_133.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1657\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_134.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1637\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_135.001 has 1 alignments\n",
      "penA length: 1755\n",
      "Alignment length: 1755\n",
      "Number of identities: 1631\n",
      "Number of gaps: 9\n",
      "\n",
      "penA_136.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1571\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_137.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_138.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1666\n",
      "Number of gaps: 3\n",
      "\n",
      "penA_139.001 has 1 alignments\n",
      "penA length: 1751\n",
      "Alignment length: 1751\n",
      "Number of identities: 1589\n",
      "Number of gaps: 7\n",
      "\n",
      "penA_140.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1538\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_141.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1595\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_142.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_143.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1579\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_144.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1592\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_145.001 has 1 alignments\n",
      "penA length: 1749\n",
      "Alignment length: 1749\n",
      "Number of identities: 1674\n",
      "Number of gaps: 0\n",
      "\n",
      "penA_146.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1585\n",
      "Number of gaps: 11\n",
      "\n",
      "penA_147.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1578\n",
      "Number of gaps: 14\n",
      "\n",
      "penA_148.001 has 1 alignments\n",
      "penA length: 1756\n",
      "Alignment length: 1756\n",
      "Number of identities: 1594\n",
      "Number of gaps: 11\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for record in blast_records:\n",
    "    print('%s has %s alignments'%(record.query, len(record.alignments)))\n",
    "    if len(record.alignments)>0:\n",
    "        for alignment in record.alignments:\n",
    "            for hsp in alignment.hsps:\n",
    "                print('penA length: %s'%len(hsp.query))\n",
    "                print('Alignment length: %s'%hsp.align_length)\n",
    "                print('Number of identities: %s'%hsp.identities)\n",
    "                print('Number of gaps: %s\\n'%hsp.gaps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 2\n",
    "You will get multiple hits, you need to define which hit is the closest match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['penA_60.001', 1749, 1749, 0, 'ATGTTGATTAAAAGCGAATATAAGCCCCGGATGCTGCCCAAAGAAGAGCAGGTCAAAAAGCCGATGACCAGTAACGGACGGATTAGCTTCGTCCTGATGGCAATGGCGGTCTTGTTTGCCTGTCTGATTGCCCGCGGGCTGTATCTGCAGACGGTAACGTATAACTTTTTGAAAGAACAGGGCGACAACCGGATTGTGCGGACTCAAGCATTGCCGGCTACACGCGGTACGGTTTCGGACCGGAACGGTGCGGTTTTGGCGTTGAGCGCGCCGACGGAGTCCCTGTTTGCCGTGCCTAAAGATATGAAGGAAATGCCGTCTGCCGCCCAATTGGAACGCCTGTCCGAGCTTGTCGATGTGCCGGTCGATGTTTTGAGGAACAAACTCGAACAGAAAGGCAAGTCGTTTATTTGGATCAAGCGGCAGCTCGATCCCAAGGTTGCCGAAGAGGTCAAAGCCTTGGGTTTGGAAAACTTTGTATTTGAAAAAGAATTAAAACGCCATTACCCGATGGGCAACCTGTTTGCACACGTCATCGGATTTACCGATATTGACGGCAAAGGTCAGGAAGGTTTGGAACTTTCGCTTGAAGACAGCCTGTATGGCGAAGACGGCGCGGAAGTTGTTTTGCGGGACCGGCAGGGCAATATTGTGGACAGCTTGGACTCCCCGCGCAATAAAGCACCGCAAAACGGCAAAGACATCATCCTTTCCCTCGATCAGAGGATTCAGACCTTGGCCTATGAAGAGTTGAACAAGGCGGTCGAATACCATCAGGCAAAAGCCGGAACGGTGGTGGTTTTGGATGCCCGCACGGGGGAAATCCTCGCCTTGGCCAATACGCCCGCCTACGATCCCAACAGACCCGGCCGGGCAGACAGCGAACAGCGGCGCAACCGTGCCGTTACCGACATGATCGAACCTGGTTCTGTCATGAAGCCGTTTACCATTGCCAAAGCATTGGATTCAGGCAAAGTGGATCCAACAGACACATTCAATACCCTGCCTTACAAAATCGGTCCGGCTACCGTACAAGATACCCACGTTTATCCTACTTTGGATGTGCGCGGCATTATGCAAAAATCTTCCAACGTCGGTACCAGTAAACTTTCTGCCATGTTTACGCCTAAAGAAATGTACGATTTCTATCACGATTTAGGTGTGGGCGTGCGCATGCATTCAGGCTTTCCCGGCGAGACTGCAGGTTTGTTGAGAAGCTGGCGCAGATGGCAAAAAATCGAACAGGCAACCATGTCTTTCGGTTATGGCCTGCAATTAAGCCTGTTGCAATTGGCGCGTGCCTATACTGTCTTGACCCATGACGGCGAATTGTTGCCGGTCAGCTTTGAAAAACAAGCAGTTGCCCCTAAAGGCAAACGCGTCATCAAAGCCTCTACTGCAAAAAAAGTGCGCGAGTTGATGGTTTCCGTTACTGAAGCCGGCGGTAGCGGTATTGCCGGTGCGGTAGATGGTTTCGACGTCGGCGCAAAAACCGGTACGGCGCGTAAGTTGGTTAACGGTCGTTACGTCGATTACAAACACGTTGCCACTTTCATCGGTTTTGCCCCGGCTAAAAATCCGCGTGTGATTGTGGCGGTAACCATTGACGAGCCGACTGCAAACGGTTACTACAGCGGCGTAGTGACAGGTCCGGTCTTCAAACAAGTTATGGGCGGTAGCCTGAACATCTTGGGCGTTTCTCCGACCAAACCTCTGACCAATGTTGCAGCCGTCAAAACACCGTCTTAA']\n"
     ]
    }
   ],
   "source": [
    "hsp_summary_best = None\n",
    "\n",
    "# iterate through\n",
    "for record in blast_records:\n",
    "    if len(record.alignments)>0:\n",
    "        for alignment in record.alignments:\n",
    "            for hsp in alignment.hsps:\n",
    "                \n",
    "                # initialize best\n",
    "                if not hsp_summary_best:\n",
    "                    hsp_summary_best = [record.query, hsp.align_length, hsp.identities, hsp.gaps, hsp.sbjct]\n",
    "                \n",
    "                # record current\n",
    "                hsp_summary_current = [record.query, hsp.align_length, hsp.identities, hsp.gaps, hsp.sbjct]\n",
    "                \n",
    "                # update best if current is more alligned\n",
    "                if hsp_summary_current[2] > hsp_summary_best[2]:\n",
    "                    hsp_summary_best = hsp_summary_current\n",
    "\n",
    "print(hsp_summary_best)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tip 3\n",
    "Once you find which hit is the closest match, check to see if there is an exact match, and translate the sequence to see which of these amino acid substitutions is present - A311V, T316P and T483S (for this last part you may wish to turn the `hsp.sbjct` text from the BLAST output back into a BioPython sequence object - this will help you - https://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html).\n",
    "\n",
    "- A311V\n",
    "- T316P\n",
    "- T483S "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: <unknown id>\n",
      "Name: penA_60.001\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGTTGATTAAAAGCGAATATAAGCCCCGGATGCTGCCCAAAGAAGAGCAGGTC...TAA')\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "seq = SeqRecord(Seq(hsp_summary_best[4]), name=hsp_summary_best[0])\n",
    "print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "V\n",
      "T\n",
      "S\n"
     ]
    }
   ],
   "source": [
    "protein = seq.translate()\n",
    "print(protein[310], protein[315], protein[482], sep='\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- A311**V**: present\n",
    "- T316**P**: absent\n",
    "- T483**S**: present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
